ratpack-test/src/main/java/ratpack/test/http/internal/DefaultMultipartForm.java

```diff
@@ -0,0 +1,396 @@
+          addSingleFileToField(filesByField, out);
+        }
+      });
+    }
+
+    private void addMultipleFilesToField(Set<File> files, String name, StringBuilder out) {
+      String subBoundary = generateSubBoundary(data.boundary, name);
+
+      declareField(name, out);
+      type(String.format("multipart/mixed, boundary=%s\n", subBoundary), out);
+      files.forEach((file) -> {
+        openBoundary(subBoundary, out);
+        disposition(String.format("attachment; filename=\"%s\"", file.getName()), out);
+        file(file, out);
+      });
+      closeBoundary(subBoundary, out);
+    }
+
+    private void addSingleFileToField(Set<File> files, StringBuilder out) {
+      File file = files.stream().findFirst().get();
+      disposition(String.format("form-data; name=\"%s\"; filename=\"%s\"",
+        file.getField(), file.getName()), out);
+      file(file, out);
+    }
+
+    private void data(byte[] value, StringBuilder out) {
+      out.append("\n");
+      out.append(new String(value));
+      out.append("\n");
+    }
+
+    private void declareField(String name, StringBuilder out) {
+      disposition(String.format("form-data; name=\"%s\"", name), out);
+    }
+
+    private void disposition(String value, StringBuilder out) {
+      label("Content-Disposition", value, out);
+    }
+
+    private void encoding(String value, StringBuilder out) {
+      if (value != null && !value.isEmpty()) {
+        label("Content-Transfer-Encoding", value, out);
+      }
+    }
+
+    private void file(File file, StringBuilder out) {
+      type(file.getContentType(), out);
+      encoding(file.getEncoding(), out);
+      data(file.getData(), out);
+    }
+
+    private String generateSubBoundary(String boundary, String name) {
+      return String.format("%s_%s", boundary, name);
+    }
+
+    private void type(String value, StringBuilder out) {
+      label("Content-Type", value, out);
+    }
+
+    private void label(String label, String value, StringBuilder out) {
+      out.append(label);
+      out.append(": ");
+      out.append(value);
+      out.append("\n");
+    }
+
+    private void openBoundary(String boundary, StringBuilder out) {
+      out.append("--");
+      out.append(boundary);
+      out.append("\n");
+    }
+
+    private void closeBoundary(String boundary, StringBuilder out) {
+      out.append("--");
+      out.append(boundary);
+      out.append("--\n");
+    }
+
+  }
+
+  final static class Data {
+
+    final String boundary;
+    final Map<String, Set<String>> fields;
+    final Map<String, Set<File>> files;
+
+    Data(String boundary, Map<String, Set<String>> fields, Map<String, Set<File>> files) {
+      this.boundary = boundary;
+      this.fields = Collections.unmodifiableMap(fields);
+      this.files = Collections.unmodifiableMap(files);
+    }
+
+  }
+
+  public final static class Builder implements MultipartFormSpec {
+
+    private final String _boundary;
+    private final Map<String, Set<String>> _fields;
```

Comment:
Why is this a map to a set? Is there anything in the spec for `multipart/from-data` that prevents you from having the same value multiple times for a single field name? Because this implementation does.